### PR TITLE
omniauth wrapper

### PR DESCRIPTION
**Pull request by pedrocarmona**
### Use omniauth hash, because allows re-use the implementation methods of find/create/update user. 

Allows to use existing omniauth oauth2 clients such as
- omniauth-facebook
- omniauth-google-oauth2

Usage: (please see code for examples)
- Either directly, by inserting all the required configs, such as client id, secret, and options. 
- Or automagically loading the devise configuration

### CODE REVIEW

**Comment by pedrocarmona** [comment]
Still need to add tests, and there might be some bugs here :)

File: lib/doorkeeper/grants_assertion/devise/omniauth.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+              client_secret: client_secret,
+              client_options: client_options,
+              assertion: assertion
+            )
+          end
+        end
+      end
+    end
+  end
+end
```
[the remark this answers is not in the document]
Comment (pedrocarmona) [reply lib/doorkeeper/grants_assertion/devise/omniauth.rb]: fixed

File: lib/doorkeeper/grants_assertion/omniauth.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+        def wrapper_klass(strategy_class)
+          Class.new(strategy_class) do
+            attr_reader :access_token
+          end
+        end
+      end
+    end
+  end
+end
```
[the remark this answers is not in the document]
Comment (pedrocarmona) [reply lib/doorkeeper/grants_assertion/omniauth.rb]: fixed

File: lib/doorkeeper/grants_assertion/devise/omniauth.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
+module Doorkeeper
```
[the remark this answers is not in the document]
Comment (pedrocarmona) [reply lib/doorkeeper/grants_assertion/devise/omniauth.rb]: fixed

File: lib/doorkeeper/grants_assertion/omniauth.rb [hunk 4]
```diff
@@ -0,0 +1,53 @@
+module Doorkeeper
```
[the remark this answers is not in the document]
Comment (pedrocarmona) [reply lib/doorkeeper/grants_assertion/omniauth.rb]: fixed

**Comment by pedrocarmona** [comment]
@matfiz please give feedback

**Comment by matfiz** [comment]
@pedrocarmona Great work! I will code review it tonight

File: lib/doorkeeper/grants_assertion/omniauth.rb [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+          client_secret:,
+          client_options:,
+          assertion:,
+          expires_at: nil,
+          expires_in: nil,
+          refresh_token: nil
+        )
+          app = nil # strategy_class is a rack middleware
+          args = [client_id, client_secret, client_options]
+          wrapper = Class.new(strategy_class).new(app, *args)
```
Comment (pedrocarmona) [inline lib/doorkeeper/grants_assertion/omniauth.rb]: anonymous class :)

**Comment by matfiz** [comment]
@pedrocarmona you code looks ok, although I haven't tested it in real app. Generally you have an excellent idea, thanks for this PR! Would you be so kind to add some info to readme? Or just drop a short info there and provide more info (with the examples taken from your code) in the wiki.

I would also really much appreciate some tests, so we can be sure it will work.

**Comment by matfiz** [comment]
@pedrocarmona I appreciate your work. Do you think you will have time to write some tests?

**Comment by pedrocarmona** [comment]
Hi @matfiz, i will write them today

**Comment by pedrocarmona** [comment]
had to update test libraries, will finish tests tomorrow

File: spec/support/dependencies/factory_bot.rb [hunk 6]
```diff
@@ -0,0 +1,2 @@
+require 'factory_bot'
```
[the remark this answers is not in the document]
Comment (pedrocarmona) [reply spec/support/dependencies/factory_bot.rb]: done

File: spec/dummy/config/initializers/doorkeeper.rb [hunk 7]
```diff
@@ -33,10 +46,10 @@
   # use_refresh_token
 
   # Provide support for an owner to be assigned to each registered application (disabled by default)
-  # Optional parameter :confirmation => true (default false) if you want to enforce ownership of
+  # Optional parameter confirmation: true (default false) if you want to enforce ownership of
```
[the remark this answers is not in the document]
Comment (pedrocarmona) [reply spec/dummy/config/initializers/doorkeeper.rb]: ignore (generated by doorkeeper install)

File: spec/doorkeeper/grants_assertion/omniauth_spec.rb [hunk 8]
```diff
@@ -0,0 +1,74 @@
+require 'spec_helper'
```
[the remark this answers is not in the document]
Comment (pedrocarmona) [reply spec/doorkeeper/grants_assertion/omniauth_spec.rb]: done

**Comment by v-kumar** [comment]
Any updates on this work?

**Comment by matfiz** [comment]
@pedrocarmona will you have time soon to apply last fixes so CI passes?

**Comment by pedrocarmona** [comment]
Hi Guys, sorry for delay, yes will do it tonight

File: spec/dummy/db/schema.rb [hunk 9]
```diff
@@ -11,9 +11,9 @@
 #
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema.define(version: 20130902175349) do
+ActiveRecord::Schema.define(version: 20180720202342) do
```
[the remark this answers is not in the document]
Comment (pedrocarmona) [reply spec/dummy/db/schema.rb]: ignore

File: spec/dummy/config/initializers/doorkeeper.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (pedrocarmona) [reply spec/dummy/config/initializers/doorkeeper.rb]: ignore

File: spec/dummy/config/initializers/devise.rb [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+  # If you want to use other strategies, that are not supported by Devise, or
+  # change the failure app, you can configure them inside the config.warden block.
+  #
+  # config.warden do |manager|
+  #   manager.intercept_401 = false
+  #   manager.default_strategies(scope: :user).unshift :some_external_strategy
+  # end
+
+  # ==> Mountable engine configurations
+  # When using Devise inside an engine, let's call it `MyEngine`, and this engine
```
[the remark this answers is not in the document]
Comment (pedrocarmona) [reply spec/dummy/config/initializers/devise.rb]: ignore

File: spec/spec_helper.rb [hunk 11]
```diff
@@ -1,2 +1,162 @@
...
+        headers: {
+         "Accept" => "*/*",
+         "Accept-Encoding" => "gzip;q=1.0,deflate;q=0.6,identity;q=0.3",
+         "Authorization" => "Bearer #{GOOGLE_VALID_TOKEN}",
+         "User-Agent" => /Faraday/
+        }
+      ).to_return(
+        status: 200,
+        body: (
+          '''
```
[the remark this answers is not in the document]
Comment (pedrocarmona) [reply spec/spec_helper.rb]: ignore

**Comment by pedrocarmona** [comment]
Hi @matfiz, sorry for delay. Here it is now with the tests!
All tests are green.
Updated a lot of stuff.
In my last commit had to remove heredocs, and the hound created yet another style comments that I ignored.

can you also remove the description of this git repo, right now it says:
```
IN NEED OF MAINTAINER.
```

But we are maintaining it, so we should remove that

**Comment by matfiz** [comment]
Thanks SO MUCH, @pedrocarmona! You've done great work creating OmniAuth wrapper and documenting it properly in the readme, so new users can easily adopt it. Such a great improvement for this little gem! Kudos for you. 🎆

**Comment by pedrocarmona** [comment]
Hi @matfiz, Thank you very much. One last thing, we need to release in rubygems:
https://rubygems.org/gems/doorkeeper-grants_assertion

**Comment by matfiz** [comment]
Yes, yes, I'm on it :)

**Comment by v-kumar** [comment]
Any update on the gem release?

**Comment by matfiz** [comment]
@v-kumar it has been released https://rubygems.org/gems/doorkeeper-grants_assertion/versions/0.2.0 :wink:

**Pull request merged**